File: src/main/java/com/currencies/mainpackage/api/controllers/TicketController.java
package com.currencies.mainpackage.api.controllers;

import com.currencies.mainpackage.models.CreateTicketRequest;
import com.currencies.mainpackage.entities.Ticket;
import com.currencies.mainpackage.app.service.TicketService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;

import java.sql.Date;
import java.util.List;

import static org.springframework.http.MediaType.*;

@RestController
@RequestMapping("/api/v1/tickets")
@RequiredArgsConstructor
public class TicketController {
    private final TicketService ticketService;

    @GetMapping(produces = APPLICATION_JSON_VALUE)
    public List<Ticket> findAll() {
        return ticketService.findAll();
    }

    @GetMapping(value = "/identifiers/{id}", produces = APPLICATION_JSON_VALUE)
    public Ticket findById(@PathVariable Integer id) {
        return ticketService.findById(id);
    }

    @GetMapping(value = "/search", produces = APPLICATION_JSON_VALUE)
    public List<Ticket> findByFromAndFromTo(@RequestParam String fromPlace,
                                                    @RequestParam String toPlace,
                                                    @RequestParam Date when,
                                                    Model model) {
//        model.addAttribute("fromPlace", fromPlace)
//                .addAttribute("toPlace", toPlace)
//                .addAttribute("when", when);
        return ticketService.findTickets(fromPlace, toPlace, when);
    }

    @PostMapping(value = "/", consumes = APPLICATION_JSON_VALUE, produces = APPLICATION_JSON_VALUE)
    public Ticket create(@RequestBody CreateTicketRequest request) {
        return ticketService.createTicket(request);
    }

    @PatchMapping(value = "/identifiers/{id}", consumes = APPLICATION_JSON_VALUE, produces = APPLICATION_JSON_VALUE)
    public Ticket update(@PathVariable Integer id, @RequestBody CreateTicketRequest request) {
        return ticketService.update(id, request);
    }

    @ResponseStatus(HttpStatus.NO_CONTENT)
    @DeleteMapping(value = "/identifiers/{id}", produces = APPLICATION_JSON_VALUE)
    public void delete(@PathVariable Integer id) {
        ticketService.delete(id);
    }

}

File: src/main/java/com/currencies/mainpackage/api/controllers/TicketController.kt
package com.currencies.mainpackage.api.controllers

import com.currencies.mainpackage.api.ApiPath.ID
import com.currencies.mainpackage.api.ApiPath.PURCHASE
import com.currencies.mainpackage.api.ApiPath.SEARCH
import com.currencies.mainpackage.api.ApiPath.TICKETS
import com.currencies.mainpackage.api.docs.SwaggerTicketController
import com.currencies.mainpackage.api.dto.request.CreateTicketRequest
import com.currencies.mainpackage.api.dto.response.TicketResponse
import com.currencies.mainpackage.core.ticket.TicketService
import com.currencies.mainpackage.entities.Ticket
import java.sql.Date
import java.time.Duration
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam

@Controller
@CrossOrigin
@RequestMapping(TICKETS)
class TicketController(private val ticketService: TicketService) : SwaggerTicketController {

    @PreAuthorize("hasRole('ROLE_ADMIN')")
    @GetMapping
    override fun findAll(): ResponseEntity<List<TicketResponse>> {
        val tickets = ticketService.findAll()
        return ResponseEntity.ok().body(tickets)
    }

    @PreAuthorize("hasRole('ROLE_ADMIN')")
    @GetMapping(ID)
    override fun findById(@PathVariable id: Long): ResponseEntity<TicketResponse> {
        return ResponseEntity.ok().body(ticketService.findById(id))
    }

    @GetMapping(SEARCH)
    fun findByFromAndTo(
        @RequestParam fromPlace: String,
        @RequestParam toPlace: String,
        @RequestParam `when`: Date,
        model: Model
    ): String {
        val ticketRows = ticketService.findTickets(fromPlace, toPlace, `when`)
        model.addAttribute("ticketRows", ticketRows)
        model.addAttribute("ticketPurchaseUrl", PURCHASE)
        model.addAttribute("isPurchaseButtonShown", true)

        return "fragments/ticketsList :: ticketsTable"
    }

    @PreAuthorize("hasRole('ROLE_ADMIN')")
    @PostMapping
    override fun create(@RequestBody request: CreateTicketRequest): ResponseEntity<TicketResponse> {
        return ResponseEntity.ok().body(ticketService.save(request))
    }

    @PreAuthorize("hasRole('ROLE_ADMIN')")
    @PatchMapping(ID)
    override fun update(@PathVariable id: Long, @RequestBody request: CreateTicketRequest): ResponseEntity<TicketResponse> {
        return ResponseEntity.ok().body(ticketService.update(id, request))
    }

    @PreAuthorize("hasRole('ROLE_ADMIN')")
    @DeleteMapping(ID)
    override fun delete(@PathVariable id: Long): ResponseEntity<Void> {
        ticketService.delete(id)
        return ResponseEntity.noContent().build()
    }

}
